### DISCUSSION TITLE
A new renderer design for web and next-gen

### DISCUSSION THREAD

**Discussion by ykafia** [stride3d/stride#2185] [Ideas]
## A bit of info 

Looking at how games are released these days we can kinda see a trend with game developers :
- Gacha/Casual games are typically released on smartphones are using very basic features, devs usually focus on making the game look smooth, some of those games end up on mid to high end devices like consoles or desktop platforms.
- Small games usually target a wide range of devices, low-end to high-end devices.
- Big games target high-end devices on launch and tend to receive patches making them run on mid-range devices.

## Graphics profile level

Our C# and shader code base rely on the [graphics profile level classification](https://github.com/stride3d/stride/blob/ec3a6e8b228bcb0211734307be1455926055ea89/sources/engine/Stride/Graphics/GraphicsProfile.cs). Most of these levels are outdated since [70% of desktops in the world are using windows](https://gs.statcounter.com/os-market-share/desktop/worldwide) to which [95% of them are using win10 and win11](https://gs.statcounter.com/os-version-market-share/windows/desktop/worldwide), confirming 95% of desktop windows users can run DX11 for games. 

Smartphone/embedded systems are where this feature would have been the most useful where the device power can vary a lot from SBCs like Raspberry Pi to higher end smartphones and retro consoles like AYN Odin 2, or Retroid.

On the subject of next-gen consoles, aside from the Nintendo Switch they all use next-gen APIs similar to DX12 or Vulkan.

## A new classification

We can end up classifying graphics profile with only 3 levels :
1. Next-Gen : The renderer would be using the next-gen apis, be fully data-driven without regards to what low end device support. This would be written in Vulkan and DirectX 12 and could use Mesh/Task shaders as well as Ray tracing pipelines.
3. Wide support : This renderer would make use of WebGPU api and some of its extensions. WebGPU is supposed to work on all devices. Stride should feature a renderer using WebGPU specifically. WebGPU provide support for only Vertex and Fragment shaders.
5. Web : This can be done through either WebGL or WebGPU

## Considerations

This implies a lot of breaking changes, and i think would potentially affect the asset system, the shader compiler as well as the editor.

**Comment by Doprez** [comment]
adding Jkaos PR here since discussions dont seem to link when it was referenced. [Drop support for DirectX 9 / OpenGL 2.0](https://github.com/stride3d/stride/pull/2186)
